Assets/MixedRealityToolkit/InputModule/Scripts/Utilities/Interactions/TwoHandManipulatable.cs

```diff
@@ -0,0 +1,631 @@
+{
+    /// <summary>
+    /// This script allows for an object to be movable, scalable, and rotatable with one or two hands. 
+    /// You may also configure the script on only enable certain manipulations. The script works with 
+    /// both HoloLens' gesture input and immersive headset's motion controller input.
+    /// See Assets/MixedRealityToolkit-Examples/Input/Readme/README_TwoHandManipulationTest.md
+    /// for instructions on how to use the script.
+    /// </summary>
+    public class TwoHandManipulatable : MonoBehaviour, IInputHandler, ISourceStateHandler
+    {
+        // Event that gets raised when user begins manipulating the object
+        public event Action StartedManipulating;
+        // Event that gets raised when the user ends manipulation
+        public event Action StoppedManipulating;
+        // Bounding Box visual on manipulation interaction
+        public GameObject GlobalBoundingBox;
+
+        [Tooltip("Transform that will be dragged. Defaults to the object of the component.")]
+        public Transform HostTransform;
+
+        public enum TwoHandedManipulation
+        {
+            Scale,
+            Rotate,
+            MoveScale,
+            RotateScale,
+            MoveRotateScale
+        };
+
+        [Tooltip("What manipulation will two hands perform?")]
+        public TwoHandedManipulation ManipulationMode;
+
+        [Tooltip("Constrain rotation along an axis")]
+        public HandlebarRotateLogic.RotationConstraint ConstraintOnRotation = HandlebarRotateLogic.RotationConstraint.None;
+        [Tooltip("If true, grabbing the object with one hand will initiate movement.")]
+        public bool OneHandMovement = true;
+
+        [Flags]
+        private enum State
+        {
+            Start = 0x000,
+            Moving = 0x001,
+            Scaling = 0x010,
+            Rotating = 0x100,
+            MovingScaling = 0x011,
+            RotatingScaling = 0x110,
+            MovingRotatingScaling = 0x111
+        };
+
+        private State currentState;
+        private MoveSphericalCoordsLogic m_moveLogic;
+        private ScaleLogic m_scaleLogic;
+        private HandlebarRotateLogic m_rotateLogic;
+        // Maps input id -> position of hand
+        private readonly Dictionary<uint, Vector3> m_handsPressedLocationsMap = new Dictionary<uint, Vector3>();
+        // Maps input id -> input source. Then obtain position of input source using currentInputSource.TryGetGripPosition(currentInputSourceId, out inputPosition);
+        private readonly Dictionary<uint, IInputSource> m_handsPressedInputSourceMap = new Dictionary<uint, IInputSource>();
+
+        private void Awake()
+        {
+            m_moveLogic = new MoveSphericalCoordsLogic();
+            m_rotateLogic = new HandlebarRotateLogic(ConstraintOnRotation);
+            m_scaleLogic = new ScaleLogic();
+
+            GlobalBoundingBox = GameObject.Find("BoundingBoxBasic");
```

Comment:
Where is this object located in relation to this GameObject the component is on?

Reply:
@cre8ivepark should comment more here, but my understanding is that this 'BoundingBoxBasic" UI element can be located anywhere in the scene. Actually after reading the code, we may want to just have a separate component "ShowBoundingBoxWhenManipulating" that would listen to the manipulation started and stopped events, and show/hide the bounding box. That would separate the concerns of two handed manipulation (which just manipulates objects) and the rendering of feedback (showing the bounding box). I'll see if I can do this over the weekend, or Yoon/somebody else please feel free to make the change.

Reply:
Just wondering if we can assign the GameObject in the inspector and maybe provide a way to search if it's unset.

Reply:
@johnppella

Reply:
changed search for GlobalBoundingBox in scene to Instatiation of BoundingBox prefab.